File: web/lib/types/copilot.ts
export type CopilotTool = {
  name: string;
  title: string;
  description: string;
  intent_family: string;
  mutates_state: boolean;
  availability: "available" | "planned";
};

export type CopilotValidationResult = {
  ok: boolean;
  code: string;
  message: string;
};

export type CopilotOpenShiftResultItem = {
  shift_id: string;
  location_id?: string | null;
  location_name?: string | null;
  role_name?: string | null;
  starts_at: string;
  ends_at: string;
  status: string;
};

export type CopilotOpenShiftsResult = {
  kind: "open_shifts";
  total_open_shifts: number;
  location_count?: number;
  items: CopilotOpenShiftResultItem[];
};

export type CopilotCampaignResultItem = {
  campaign_id: string;
  shift_id: string;
  location_name?: string | null;
  role_name?: string | null;
  status: string;
  phase_target?: string | null;
  opened_at?: string | null;
};

export type CopilotCampaignsResult = {
  kind: "campaigns";
  total_active_campaigns: number;
  running_count?: number;
  queued_count?: number;
  items: CopilotCampaignResultItem[];
};

export type CopilotManagerActionResultItem = {
  campaign_id: string;
  shift_id: string;
  location_name?: string | null;
  role_name?: string | null;
  starts_at?: string | null;
  status: string;
};

export type CopilotManagerActionsResult = {
  kind: "manager_actions";
  total_actions: number;
  items: CopilotManagerActionResultItem[];
};

export type CopilotAvailabilityRuleResultItem = {
  day_of_week: number;
  start_local_time: string;
  end_local_time: string;
  timezone: string;
};

export type CopilotAvailabilityUpdateResult = {
  kind: "availability_update";
  employee_id: string;
  employee_name: string;
  timezone: string;
  day_count: number;
  rule_count: number;
  rules: CopilotAvailabilityRuleResultItem[];
};

export type CopilotHelpResult = {
  kind: "help";
  tools: CopilotTool[];
};

export type CopilotToolResultPayload =
  | CopilotOpenShiftsResult
  | CopilotCampaignsResult
  | CopilotManagerActionsResult
  | CopilotAvailabilityUpdateResult
  | CopilotHelpResult
  | Record<string, unknown>;

export type CopilotIntent = {
  family: string;
  tool_name: string;
  reasoning: string;
  confidence: number;
};

export type CopilotSession = {
  id: string;
  business_id: string;
  location_id?: string | null;
  operator_user_id: string;
  channel_last_seen: string;
  intent_family?: string | null;
  state:
    | "active"
    | "awaiting_confirmation"
    | "awaiting_clarification"
    | "completed"
    | "expired";
  context_profile: Record<string, unknown>;
  working_memory: Record<string, unknown>;
  expires_at?: string | null;
  last_message_at?: string | null;
  created_at: string;
};

export type CopilotMessage = {
  id: string;
  copilot_session_id: string;
  direction: "inbound" | "outbound";
  normalized_channel: string;
  raw_text: string;
  normalized_text: string;
  message_metadata: Record<string, unknown>;
  created_at: string;
};

export type CopilotActionRun = {
  id: string;
  copilot_session_id: string;
  tool_name: string;
  status: "planned" | "validated" | "executed" | "failed" | "cancelled";
  input_payload: Record<string, unknown>;
  validation_result: CopilotValidationResult;
  result_payload: CopilotToolResultPayload;
  error_payload: Record<string, unknown>;
  started_at: string;
  finished_at?: string | null;
};

export type CopilotSessionDetail = {
  session: CopilotSession;
  tools: CopilotTool[];
  messages: CopilotMessage[];
  action_runs: CopilotActionRun[];
};

export type CopilotTurn = {
  session: CopilotSession;
  resolved_intent: CopilotIntent;
  inbound_message: CopilotMessage;
  outbound_message: CopilotMessage;
  action_run: CopilotActionRun;
  tools: CopilotTool[];
};

export type CopilotSessionCreatePayload = {
  location_id?: string | null;
  normalized_channel?: string;
  reuse_active?: boolean;
};

export type CopilotMessageCreatePayload = {
  text: string;
  location_id?: string | null;
  normalized_channel?: string;
};

export type CopilotLiveProgressState =
  | "thinking"
  | "planning"
  | "running_tool";

export type CopilotLiveEvent =
  | {
      event_id: string;
      event_type: "session.ready";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { detail: CopilotSessionDetail };
    }
  | {
      event_id: string;
      event_type: "user.message.accepted";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { message: CopilotMessage; session: CopilotSession };
    }
  | {
      event_id: string;
      event_type: "assistant.turn.started";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { session: CopilotSession; message_id: string };
    }
  | {
      event_id: string;
      event_type: "assistant.progress";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: {
        state: CopilotLiveProgressState;
        label: string;
        resolved_intent?: CopilotIntent;
        planner_source?: string;
      };
    }
  | {
      event_id: string;
      event_type: "tool.started";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: {
        tool_name: string;
        input_payload: Record<string, unknown>;
        validation_result: CopilotValidationResult;
      };
    }
  | {
      event_id: string;
      event_type: "tool.finished";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { action_run: CopilotActionRun };
    }
  | {
      event_id: string;
      event_type: "assistant.message.completed";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { turn: CopilotTurn };
    }
  | {
      event_id: string;
      event_type: "assistant.turn.failed";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: {
        error: { code: string; message: string };
        turn?: CopilotTurn;
      };
    }
  | {
      event_id: string;
      event_type: "session.error";
      trace_id: string;
      session_id: string;
      occurred_at: string;
      payload: { code: string; message: string };
    };

File: app/schemas/copilot.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import Field

from app.schemas.common import BaseSchema


class CopilotToolRead(BaseSchema):
    name: str
    title: str
    description: str
    intent_family: str
    mutates_state: bool = False
    availability: Literal["available", "planned"] = "available"


class CopilotSessionCreate(BaseSchema):
    location_id: Optional[UUID] = None
    normalized_channel: str = "dashboard"
    reuse_active: bool = True


class CopilotValidationResultRead(BaseSchema):
    ok: bool
    code: str
    message: str


class CopilotIntentRead(BaseSchema):
    family: str
    tool_name: str
    reasoning: str
    confidence: float = Field(ge=0.0, le=1.0)


class CopilotSessionRead(BaseSchema):
    id: UUID
    business_id: UUID
    location_id: Optional[UUID] = None
    operator_user_id: UUID
    channel_last_seen: str = "dashboard"
    intent_family: Optional[str] = None
    state: Literal[
        "active",
        "awaiting_confirmation",
        "awaiting_clarification",
        "completed",
        "expired",
    ] = "active"
    context_profile: dict = Field(default_factory=dict)
    working_memory: dict = Field(default_factory=dict)
    expires_at: Optional[datetime] = None
    last_message_at: Optional[datetime] = None
    created_at: datetime


class CopilotMessageRead(BaseSchema):
    id: UUID
    copilot_session_id: UUID
    direction: Literal["inbound", "outbound"]
    normalized_channel: str = "dashboard"
    raw_text: str
    normalized_text: str
    message_metadata: dict = Field(default_factory=dict)
    created_at: datetime


class CopilotActionRunRead(BaseSchema):
    id: UUID
    copilot_session_id: UUID
    tool_name: str
    status: Literal["planned", "validated", "executed", "failed", "cancelled"]
    input_payload: dict = Field(default_factory=dict)
    validation_result: CopilotValidationResultRead
    result_payload: dict = Field(default_factory=dict)
    error_payload: dict = Field(default_factory=dict)
    started_at: datetime
    finished_at: Optional[datetime] = None


class CopilotSessionDetailRead(BaseSchema):
    session: CopilotSessionRead
    tools: list[CopilotToolRead] = Field(default_factory=list)
    messages: list[CopilotMessageRead] = Field(default_factory=list)
    action_runs: list[CopilotActionRunRead] = Field(default_factory=list)


class CopilotMessageCreate(BaseSchema):
    text: str
    location_id: Optional[UUID] = None
    normalized_channel: str = "dashboard"


class CopilotTurnRead(BaseSchema):
    session: CopilotSessionRead
    resolved_intent: CopilotIntentRead
    inbound_message: CopilotMessageRead
    outbound_message: CopilotMessageRead
    action_run: CopilotActionRunRead
    tools: list[CopilotToolRead] = Field(default_factory=list)


class CopilotSessionEventRead(BaseSchema):
    event_id: UUID
    event_type: Literal[
        "session.ready",
        "user.message.accepted",
        "assistant.turn.started",
        "assistant.progress",
        "tool.started",
        "tool.finished",
        "assistant.message.completed",
        "assistant.turn.failed",
        "session.error",
    ]
    trace_id: str
    session_id: UUID
    occurred_at: datetime
    payload: dict = Field(default_factory=dict)
